mava/systems/ppo/sebulba/rec_ippo.py

```diff
@@ -0,0 +1,902 @@
+from mava.systems.ppo.types import (
+    HiddenStates,
+    OptStates,
+    Params,
+    RNNLearnerState,
+    RNNPPOTransition,
+)
+from mava.types import (
+    Metrics,
+    Observation,
+    RecActorApply,
+    RecCriticApply,
+    SebulbaLearnerFn,
+)
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.config import check_sebulba_config, check_total_timesteps
+from mava.utils.jax_utils import switch_leading_axes
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.network_utils import get_action_head
+from mava.utils.sebulba import ParamsSource, Pipeline, RecordTimeTo, ThreadLifetime
+from mava.utils.training import make_learning_rate
+from mava.wrappers.episode_metrics import get_final_step_metrics
+from mava.wrappers.gym import GymToJumanji
+
+
+def rollout(
+    key: chex.PRNGKey,
+    env: GymToJumanji,
+    config: DictConfig,
+    rollout_queue: Pipeline,
+    params_source: ParamsSource,
+    apply_fns: Tuple[RecActorApply, RecCriticApply],
+    actor_device: int,
+    seeds: List[int],
+    thread_lifetime: ThreadLifetime,
+) -> None:
+    """Runs rollouts to collect trajectories from the environment.
+
+    Args:
+        key (chex.PRNGKey): The PRNGkey.
+        config (DictConfig): Configuration settings for the environment and rollout.
+        rollout_queue (Pipeline): Queue for sending collected rollouts to the learner.
+        params_source (ParamsSource): Source for fetching the latest network parameters
+        from the learner.
+        apply_fns (Tuple): Functions for running the actor and critic networks.
+        actor_device (Device): Actor device to use for rollout.
+        seeds (List[int]): Seeds for initializing the environment.
+        thread_lifetime (ThreadLifetime): Manages the thread's lifecycle.
+    """
+    name = threading.current_thread().name
+    print(f"{Fore.BLUE}{Style.BRIGHT}Thread {name} started{Style.RESET_ALL}")
+    actor_apply_fn, critic_apply_fn = apply_fns
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def act_fn(
+        params: Params,
+        observation: Observation,
+        dones: chex.Array,
+        hstates: HiddenStates,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+
+        batched_observation = tree.map(lambda x: x[jnp.newaxis, :], observation)
+        ac_in = (
+            batched_observation,
+            dones[jnp.newaxis, :],
+        )
+        policy_hidden_state, actor_policy = actor_apply_fn(
+            params.actor_params, hstates.policy_hidden_state, ac_in
+        )
+        critic_hidden_state, value = critic_apply_fn(
+            params.critic_params, hstates.critic_hidden_state, ac_in
+        )
+
+        action = actor_policy.sample(seed=key)
+        log_prob = actor_policy.log_prob(action)
+
```

Comment:
Better to do this in the act fn, cleaner and it's jitted

```suggestion
        action = actor_policy.sample(seed=key).squeeze(0)
        log_prob = actor_policy.log_prob(action).squeeze(0)
        value = value.sqeeuze(0)

```